### PR TITLE
improve list locations

### PR DESCRIPTION
**Pull request by jrvidal**
I noticed we sometimes lack span info for `ExprKind::List` instances, while I was trying to improve error locations. So this PR does both things.

### CODE REVIEW

File: crates/steel-core/src/compiler/modules.rs [hunk 1]
```diff
@@ -2376,7 +2376,7 @@ impl<'a> ModuleBuilder<'a> {
                         match *for_syntax {
                             x if x == *FOR_SYNTAX => {
                                 if l.args.len() != 2 {
-                                    stop!(ArityMismatch => "provide expects a single identifier in the (for-syntax <ident>)")
+                                    stop!(ArityMismatch => "provide expects a single identifier in the (for-syntax <ident>)"; opt l.location)
```
Comment (jrvidal) [inline crates/steel-core/src/compiler/modules.rs:2379]: I added span info for a bunch of errors that were missing it.

File: crates/steel-core/src/rerrs.rs [hunk 2]
```diff
@@ -396,6 +396,12 @@ macro_rules! stop {
     ($type:ident => $thing:expr; $span:expr) => {
         return Err($crate::rerrs::SteelErr::new($crate::rerrs::ErrorKind::$type, ($thing).to_string()).with_span($span))
     };
+    ($type:ident => $thing:expr; opt $span:expr) => {
```
Comment (jrvidal) [inline crates/steel-core/src/rerrs.rs:399]: New variant, with `opt` indicating "optional".

File: crates/steel-parser/src/parser.rs [hunk 3]
```diff
@@ -478,9 +478,26 @@ impl<'a> Parser<'a> {
         }
     }
 
-    fn read_from_tokens(&mut self) -> Result<ExprKind> {
-        let mut stack: Vec<Vec<ExprKind>> = Vec::new();
-        let mut current_frame: Vec<ExprKind> = Vec::new();
+    fn maybe_lower_frame(&self, frame: Frame, close: Span) -> Result<ExprKind> {
```
Comment (jrvidal) [inline crates/steel-parser/src/parser.rs:481]: We can _almost_ get rid of the regular `maybe_lower`, if we short-circuit the only remaining call when parsing quotes ([here](https://github.com/mattwparas/steel/blob/842dd029978f77298768bd8f8913e3a4ef242a75/crates/steel-parser/src/parser.rs#L1022)) to generate a `ExprKind::Quote` directly.

File: src/lib.rs [hunk 4]
```diff
@@ -33,7 +33,13 @@ enum EmitAction {
     /// Output a debug display of the fully transformed bytecode
     Bytecode { default_file: Option<PathBuf> },
     /// Print a debug display of the fully expanded AST
-    Ast { default_file: Option<PathBuf> },
+    Ast {
+        default_file: Option<PathBuf>,
+        #[arg(long)]
+        expanded: Option<bool>,
+        #[arg(long)]
+        pretty: Option<bool>,
```
Comment (jrvidal) [inline src/lib.rs:41]: I took the liberty to add a couple of options to `steel ast` to help debugging.

**Comment by mattwparas** [comment]
This LGTM - any other changes you wanted to make? Otherwise I can approve and merge

**Comment by jrvidal** [comment]
Nope, this is everything I had re: list spans.

**Review by mattwparas — APPROVED** [review]

**Comment by mattwparas** [comment]
Thanks!

**Pull request merged**